### DISCUSSION TITLE
Unsaturated soil formulation

### DISCUSSION THREAD

**Discussion by Francisliii** [Anura3D/Anura3D_OpenSource#63] [Q&A] [answered]
Hi Dear all,

I am working on modeling unsaturated soil behavior using the built-in formulation "unsaturated material with 2 phase suction effects". My first model is based on the "Shallow foundation" and guided by the "Rapid drawdown" examples in the tutorial manual. The model went through eventually, but there are a few questions I would appreciate any help to clarify.

Regarding the calculation data, gravity load is added immediately in the first calculation step, and quasi-static convergences and k0 procedure are applied. 

Questions: 
1. Is hydraulic boundary condition necessary in this type of problem where transient flow is not presented? In the very first trial, I added the hydraulic boundary conditions as guided in the tutorial but the calculation returns "forrtl: severe (157): Program Exception - access violation". Then I deleted the hydraulic boundary condition and kept everything else the same and the model went through. 

Quoted from tutorial manual (11.2.1.7): "Since the body is expected to experience large displacements, initial boundary nodes may differ from boundary nodes at a later stage of the simulation, thus it is necessary to define a potential area/volume of research of boundary nodes ...". 

Since the pressing foundation caused large deformation to the unsaturated soil, deleting hydraulic boundary condition is not aligned with the quotation above. Could anyone help me understanding this part? 

2. In the case of manually adding phreatic surface to the soil, the matric suction is defined by the SWRC model (say Van Genuchten). Is there a way to maintain the matric suction value in the soil body to a relatively constant value through out the simulation? I wondered if the "initial water pressure"  option in calculation data or "2D-Liquid Pressure" option in loading conditions would do the job? 

3. What's the difference between "prescribed velocity and initial velocity". My current understanding is that the initial velocity applies in the first calculation step and the rest will be taken care of by gravity. Prescribed velocity will eventually be reached and maintained thereafter during the simulation. Is this correct?

Thank you all for any helps!
Francis

**Comment by Francisliii** [comment]
Well I figured out the third questions.

**Comment by FrancescaCeccato** [comment, marked as answer]
Dear Francis,
I'm not sure to get your problem correctly. 
1. What hydraulic boundary conditions are you using? Hydraulic head, seepage fare or infiltration? These are not always necessary, it depends on the problem you want to simulate. Remember that if you do not prescribe any condition at the boundary for the liquid pahse, zero pressure is automatically applied.
2.  the formulation is fully coupled; this means that soil deformations will modify pressures (or suctions) and pressure changes will generate displacements. It is not possible to assign a constant suction in the model in a fully coupled formulation. You may try to assign suctions at the boundaries to try keeping the suction more stable, but when the soil deforms the suction will inevitably change.
3.  prescribed velocity only applies to structures. it can be used, for example,for pile penetration, CPT, foundations ect.

**Reply by Francisliii** [reply]
Hi Dear Dr. Cecceato,

Thank you for your answers. I figured out the first and third questions by trial and error. Regarding the suction conditions in the soil model, thank you for your suggestions and I will try it.

Thanks,
Francis

**Reply by CPLA81192** [reply]
Hi Dear Dr. Cecceato,
I would like to know how to set the initial matrix suction when simulating unsaturated soil. Looking forward to your reply! Thank you!

**Reply by FrancescaCeccato** [reply]
Suction is elaborated in the model as a pressure, thus the initial suction can be applied by appropriate pressure boundary conditions or with the K0 procedure.

**Reply by CPLA81192** [reply]
Thank you for your reply. I understand how I should try. I also have another question: Can the software compute in parallel? I only have a single core CPU working during the calculation process, resulting in very low computing efficiency! Looking forward to your reply.

**Reply by FrancescaCeccato** [reply]
Yes, some parts of the code are parallelized. You can change the setting of the compiler.

**Reply by CPLA81192** [reply]
Thank you for your reply. I am trying to find a way to set it. Thank you very much.

**Reply by CPLA81192** [reply]
I'm very sorry to bother you. When simulating rainfall infiltration, when the initial liquid pressure is high, such as 2000kPa, the soil is prone to expansion and loosening, and it is easy to disconnect from the bottom. May I ask what may be the reason for this situation
![296902735-855f8934-6af1-46ed-b39e-14f8c16f2a77](https://github.com/Anura3D/Anura3D_OpenSource/assets/97551917/b9fbaa05-6b79-4bc9-bb70-e13d26fa7951)

**Reply by FrancescaCeccato** [reply]
Dear @CPLA81192,
Sorry but I do not clearly get the property of the model. Are you applying a suction at the top surface? What is the initial state?

**Reply by CPLA81192** [reply]
The soil I simulated had an initial matric suction of 2000kPa, so the INITIAL_WATER_PRESSURE of my facility was 2000kPa. However, the soil would swell during the infiltration simulation

**Reply by FrancescaCeccato** [reply]
Initial suction is 2000kPa, is it initially in equilibium with gravity? 
Then water infiltrates from the top, suction decreases and effective stress decreases, thus soil swell. If swelling is much larger than expected it may be because of the constitutive model. Probably the elastic modulus is too small and deformation due to swelling is too large.

**Reply by CPLA81192** [reply]
Dear FrancescaCeccato,
 I apologize for disturbing you once again.I am currently simulating the soil liquefaction process induced by seismic loading and rainfall infiltration.Initially,I completed the calculation of rainfall infiltration.Subsequently,I applied seismic waves through prescribed velocity inputs.However,during the computation,I observed that the entire soil mass of the slope moved upward,which I do not understand.When quasi-static convergence is enabled during the calculation,this phenomenon appears to be mitigated,but it significantly slows down the computational speed.I would like to ask for your advice on how to resolve this issue.
![微信图片_2025-03-12_092819_482](https://github.com/user-attachments/assets/a42f73e3-91b0-4535-a32c-5cd913c5cda0)
[slop_2dCPS_351.txt](https://github.com/user-attachments/files/19200917/slop_2dCPS_351.txt)
[slop_2dGOM.txt](https://github.com/user-attachments/files/19200918/slop_2dGOM.txt)

**Reply by FrancescaCeccato** [reply]
It could be a problem with the contact algorithm. Can you try withour contact just to see it the problem remains? We are working on finding a solution, but we do nto have a final outcome.

**Reply by CPLA81192** [reply]
Thank you very much for your prompt response. I am currently attempting to deactivate the contact algorithm and recompute the results.

**Reply by CPLA81192** [reply]
Dear Francesca: 
       During the simulation of infiltration, I have encountered a situation where, after initialization, if the infiltration rate is very low or infiltration is not enabled, there will be a phenomenon of water evaporation on the slope surface, which causes the suction force to increase. When I set the liquid pressure on the slope to 0, it didn't have any effect. Could you please explain what might be the reason for this phenomenon? In addition, at the bottom of the slope, although fixed boundaries were applied, the phenomenon of a decrease in suction was still observed.
![Uploading 175B1910CF31138931BA81F03E8305B3.png…]()

**Reply by FrancescaCeccato** [reply]
If there is an infiltration applied at the boundary and this is entering the system, it cannot be that water is flowing out. I suggest to check the sign of infiltration velocity. Infiltration with opposite velocity means evaporation. With zero pressure, of course water will enter an unsaturated system.
Did you fix the liquid velocity at the bottom?

**Reply by CPLA81192** [reply]
The infiltration velocity is correct, and a fixed liquid boundary is applied at the bottom. The increase in suction of the surface soil is more akin to the natural percolation of moisture within the soil under gravity, as the suction at the bottom simultaneously decreases. However, I wish to avoid this process and maintain a constant initial water content.

 Additionally, I aim to simulate a heavy rainfall infiltration process, where the surface soil approaches saturation. Yet, when I set a small permeability coefficient, the computation is very slow. Conversely, increasing the permeability coefficient leads to rapid infiltration, with the wetting front quickly penetrating the soil without saturating the surface layer, resembling a light rainfall infiltration process. How can I address this issue?

**Reply by FrancescaCeccato** [reply]
If you want to keep constant suction in the soil, you must apply proper boundary conditions for it. COnsider that the formulation is coupled, meaning that soil deformations change pressure distribution. If you want to check wether this could be the cause, you can fix the solid skeleton with the flag $FIX_SOLID_SKELETON in the CPS file. In this case only seepage is simulated.
Permeability decreases the critical time step size necessary to keep the numerical solution stable, ufortunately there is not much you can do about it. To increase the critical time step you can try to use mass scaling or reduce bulk modulus.

**Comment by Mangwuha** [comment]
Hi，there
I encountered a similar problem. I want to achieve the coupling effect of rainfall and earthquakes. Currently, I have discovered three aspects that confuse me:
1.Whether it is rain followed by an earthquake, an earthquake followed by rain, or both are activated simultaneously, when the contact algorithm is applied, there is a probability that all the particles in the model will fly up, as shown in Figure 1. When I turned off the contact algorithm, the calculation example could proceed smoothly, but the final model calculation result was definitely incorrect (such as acceleration, pore water pressure).
![1](https://github.com/user-attachments/assets/71d9434f-4ae6-4127-a1fc-0fed3b3f1279)

2. When I calculated the example of an earthquake followed by rainfall, the calculation went smoothly. However, when checking the results in Paraview, the first step after the rainfall was applied reported an error, as shown in Figure 2 below.
![2](https://github.com/user-attachments/assets/08a9c161-87e8-41cc-b2a1-1bb99264730d)

3. Under the premise that all other parameters and boundary conditions remain unchanged, I changed the grid size but obtained different pore water pressure distributions (all my calculation examples include static calculation, self-weight stress, and initial pore water pressure distribution).
![4](https://github.com/user-attachments/assets/574864b0-9010-4b77-ba96-1cae27eaaea1)
![3](https://github.com/user-attachments/assets/9855fdf4-4c06-425d-92d1-f7f9ca43e974)

May I ask if anyone can give me some suggestions? Thank you very much.

**Reply by FrancescaCeccato** [reply]
Hi!
1. there mught be problems with contact lgorithm and 2phase, we are looking into it and found some solutions although they are not fully tested. If you want to try we can explain how to modify the code.
2. it seems that one variable is not correctly calculated, maybe there is a convergence isuue
3. the two pictures that you show are clearly the results of calculations with different boundary conditions, the mesh is very similar or identical. Check initial settings, including freatic surface.

**Reply by Mangwuha** [reply]
Hello!
Thank you for your reply. I have come across some points that have confused me.
1.  When simulating earthquakes, is there a free-field boundary in Anura3D? I didn't find the relevant Settings in the GID.
2.  (Main problem)The absorbing boundary can be used to absorb seismic waves and reduce reflection. I want to know how the alpha and delta parameters in the absorbing boundary should be determined based on the parameters of the applied seismic waves?
3.  I also want to know whether the setting of the absorbing boundary can completely absorb the reflected wave?

**Reply by ayerrocolom** [reply]
Hello!
In Anura3D, we do not have the free-field boundary conditions implemented.
The absorbing boundary implemented in the code corresponds to a parallel combination of springs and dashpots. Note that the stiffness of the springs is initially assumed equal to the stiffness of the material in contact with the boundary.  The delta and alpha input parameters scale the stiffness and viscosity of these elements, respectively. The delta corresponds to the virtual thickness of the absorbing boundary (0 corresponds to a rigid boundary; infinity means the boundary behaves as a dashpot, so it will creep). On the other hand, alpha describes the resistance to deformation of the dashpot or the creep rate (large values represent a soft dashpot or high creep rate, while 0 means that the boundary behaves as a spring, no dashpot). Both parameters are problem-dependent and need to be calibrated to absorb the stress waves generated in the specific problem. This will vary with the stiffness of the materials in the model, model geometry, loading frequency.... 

I hope this helps.
Alba

**Reply by Mangwuha** [reply]
Hi
I want to try using Solid Traction A to apply seismic loads, but I don't know what the file suffix name is for using external files to control the input timeline (for example, the external file suffix of prescribed velocity is. (PVF), could you please tell me what the suffix name of the SOLID TRACTION A external file is? Thank you very much!

**Reply by ayerrocolom** [reply]
Hello,
The evolution of solid traction A can only be controlled by the load multipliers included in the CPS. The use of solid traction to apply seismic loads will need a little bit of coding.

**Reply by Mangwuha** [reply]
Hello! My model is currently encountering some issues. I apply seismic loads to a slope model and I wonder if there is any incompatibility between the application of the absorption boundary and the prescribed velocity or solid fixity boundary? I found that the absorption boundary I applied at the boundary did not work, and the internal horizontal acceleration of the slope model did not show an effect of elevation amplification. The acceleration at the top of the model is much smaller than the magnitude of the seismic load applied at the bottom. The results obtained by applying an absorption boundary are almost the same as those obtained without applying an absorption boundary. On the left and right sides of the model, I applied X solid fixity and XY solid fixity respectively. After the quasi-static process is calculated, remove the above solid fixity and apply the prescribed velocity. May I ask if the boundary conditions imposed on my model are reasonable?
![Uploading mmexport1766154404921.jpg…]()

**Reply by Mangwuha** [reply]
@ayerrocolom 
Hello! I am simulating the failure process of slopes under the action of earthquakes. After trying different ways of setting the absorption boundary, I still think my calculation results are not satisfactory, and the absorption effect of seismic waves inside the slope is not very good. Could you help me take a look at where my boundary Settings need improvement? Also, I want to know to what extent the reflected wave is absorbed to be considered good?

I have established one main model: the largest possible model (similar to a free field).To speed up the calculation efficiency, I temporarily use a two-period sine wave as a substitute.In the velocity file, approximately the first 50 seconds are sine waves, and there are no seismic waves in the last 100 seconds.

<img width="1329" height="478" alt="1" src="https://github.com/user-attachments/assets/f776aee2-0cb8-4271-93fd-56e7e5042e66" />
single point 2D model

<img width="1826" height="746" alt="2" src="https://github.com/user-attachments/assets/c5ee28d2-8a39-4d4e-b970-5e6c3e6b139d" />
solid fixities

<img width="1836" height="758" alt="3" src="https://github.com/user-attachments/assets/adf1c38f-7037-4306-9a8b-5ee506957227" />
remove fixities after applying convergence criteria

<img width="1201" height="402" alt="4" src="https://github.com/user-attachments/assets/8214327f-c887-49a6-ace5-d8ee941c9057" />
absorbing boundary applied .The left and right sides absorb seismic waves in the horizontal direction. The bottom absorbs vertical seismic waves.

prescribed velocity applied. To simulate a real earthquake, seismic waves are applied to the bottom of the model. What I'm not sure about is whether there is a contradiction among the prescribed velocity boundary, absorbing boundary and solid fixity. Is solid fixity similar to a rigid boundary that does not absorb seismic waves at all? If so, no matter what, the absorption boundary will not work. Similarly, does the problem also exist between the prescribed velocity boundary and the absorbing boundary?

The results are as follows：
<img width="941" height="655" alt="6" src="https://github.com/user-attachments/assets/428016e2-627c-4cb8-aa96-5b7cf70c92fe" />
<img width="1035" height="659" alt="7" src="https://github.com/user-attachments/assets/ddda213d-390c-4815-9c9f-36ffb23b94ca" />
<img width="945" height="655" alt="8" src="https://github.com/user-attachments/assets/82523781-194f-46e1-a9b9-f5838de9e865" />
<img width="982" height="662" alt="9" src="https://github.com/user-attachments/assets/2b16d618-f896-4a0f-a88c-3d5e7d09e295" />
It can be seen that the acceleration curve at the bottom of the model performs well, and there are no obvious reflected waves after loading. The reflected wave oscillation of the acceleration time-history curve of the monitoring points far from the bottom after 50 seconds is obvious. Except for the bottom measurement point on the far left, the acceleration of the other two measurement points is relatively small. Can it be considered that the absorption boundary is valid?

Could you guide me on how I should reasonably set these boundaries that I need to use? Thank you very much!

**Reply by Mangwuha** [reply]
@ayerrocolom 
<img width="1829" height="751" alt="5" src="https://github.com/user-attachments/assets/898db374-8d9b-4744-80c5-c5923d9cb600" />
One image was omitted. My prescribed velocity boundary Settings are as follows.

**Reply by hjk1108** [reply]
Hello, how are you doing now?

**Comment by Mangwuha** [comment]
Hi
Thank you for your reply.I will check my model.
